sw/device/lib/crypto/impl/aes_gcm/aes_gcm.c

```diff
@@ -708,3 +709,59 @@ aes_error_t aes_gcm_encrypt(const aes_key_len_t key_len,
   return aes_gcm_compute_tag(key_len, key_shares, &Htbl, plaintext_len,
                              ciphertext, aad_len, aad, &j0, tag);
 }
+
+aes_error_t aes_gcm_decrypt(const aes_key_len_t key_len,
+                            const uint32_t *key_shares[2], const size_t iv_len,
+                            const uint8_t *iv, const size_t ciphertext_len,
+                            const uint8_t *ciphertext, const size_t aad_len,
+                            const uint8_t *aad, const uint8_t *tag,
+                            uint8_t *plaintext, hardened_bool_t *success) {
+  // Check that the input parameter sizes are valid.
+  if (check_buffer_lengths(iv_len, ciphertext_len, aad_len) !=
+      kHardenedBoolTrue) {
+    return kAesInternalError;
+  }
+
+  // Compute the hash subkey H as a product table.
+  aes_gcm_product_table_t Htbl;
+  aes_error_t err = aes_gcm_hash_subkey(key_len, key_shares, &Htbl);
+  if (err != kAesOk) {
+    return err;
+  }
+
+  // Compute the counter block (called J0 in the NIST specification).
+  aes_block_t j0;
+  err = aes_gcm_counter(iv_len, iv, &Htbl, &j0);
+  if (err != kAesOk) {
+    return err;
+  }
+
+  // Compute the expected authentication tag T.
+  uint8_t expected_tag[kAesGcmTagNumBytes];
+  err = aes_gcm_compute_tag(key_len, key_shares, &Htbl, ciphertext_len,
+                            ciphertext, aad_len, aad, &j0, expected_tag);
+  if (err != kAesOk) {
+    return err;
+  }
+
+  // Copy expected and actual tag to word-size buffers to avoid violating
+  // strict aliasing rules.
```

Comment:
Nice!

If we're ever hurting for CPU cycles, I think we could eliminate these two copies. Apparently, [`hardened_memeq()`](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/base/hardened_memory.h;l=73-76;drc=96e13fe342572922d0c64b382561792d6c9b5b05) bends the strict aliasing rules similarly to `memcpy` — the only difference is that it requires 32-bit alignment. It might be possible to prove/require 32-bit alignment with `__attribute__ ((aligned (4)))` on the `tag` parameter and the `expected_tag` local.